Moodle 4.4 related deprecation warnings

Hello again.

My team is currently working on a site upgrade to Moodle 4.4, and we noticed in our CI pipeline that the Ribbons plugin is now emitting deprecation warnings. 

These are 

`Debugging: Callback before_standard_html_head in tool_ribbons component should be migrated to new hook callback for core\hook\output\before_standard_head_html_generation`

and

`Debugging: Callback before_standard_top_of_body_html in tool_ribbons component should be migrated to new hook callback for core\hook\output\before_standard_top_of_body_html_generation`

You can reproduce this by running this `mod_assign\event\events_test::test_batch_set_workflow_state_viewed` testcase in a vanilla M4.4 instance and the Ribbons plugin installed.

Example output:

```bash
stefan@nichtsnutz: ~/dev/projects/moodle44 on MOODLE_404_STABLE[?]
$ vendor/bin/phpunit  --filter=test_batch_set_workflow_state_viewed mod/assign/tests/event/events_test.php 
Moodle 4.4.3+ (Build: 20240924), 9a19ddfef471d9ccbecd753f4700a8820920aaa2
Php: 8.1.29, mysqli: 8.0.39, OS: Linux 6.8.0-45-generic x86_64
PHPUnit 9.6.18 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 00:00.841, Memory: 87.50 MB

There was 1 error:

1) mod_assign\event\events_test::test_batch_set_workflow_state_viewed
Unexpected debugging() call detected.
Debugging: Callback before_standard_html_head in tool_ribbons component should be migrated to new hook callback for core\hook\output\before_standard_head_html_generation
* line 7479 of /lib/moodlelib.php: call to debugging()
* line 7593 of /lib/moodlelib.php: call to {closure}()
* line 73 of /lib/classes/hook/output/before_standard_head_html_generation.php: call to get_plugins_with_function()
* line 718 of /lib/outputrenderers.php: call to core\hook\output\before_standard_head_html_generation->process_legacy_callbacks()
* line 219 of /lib/mustache/src/Mustache/Context.php: call to core_renderer->standard_head_html()
* line 138 of /lib/mustache/src/Mustache/Context.php: call to Mustache_Context->findVariableInStack()
* line 31 of udata/localcache/mustache/1/boost/__Mustache_73560842f9d6c5bc337558df4f13b989.php: call to Mustache_Context->findDot()
* line 13 of udata/localcache/mustache/1/boost/__Mustache_4aa1b29b14b62e9eb02cc88038b79589.php: call to __Mustache_73560842f9d6c5bc337558df4f13b989->renderInternal()
* line 67 of /lib/mustache/src/Mustache/Template.php: call to __Mustache_4aa1b29b14b62e9eb02cc88038b79589->renderInternal()
* line 216 of /lib/outputrenderers.php: call to Mustache_Template->render()
* line 107 of /theme/boost/layout/drawers.php: call to renderer_base->render_from_template()
* line 1483 of /lib/outputrenderers.php: call to include()
* line 1408 of /lib/outputrenderers.php: call to core_renderer->render_page_layout()
* line 273 of /mod/assign/classes/output/renderer.php: call to core_renderer->header()
* line 512 of /lib/outputrenderers.php: call to mod_assign\output\renderer->render_assign_header()
* line 5131 of /mod/assign/locallib.php: call to plugin_renderer_base->render()
* line 121 of /mod/assign/tests/fixtures/testable_assign.php: call to assign->view_batch_set_workflow_state()
* line 1035 of /mod/assign/tests/event/events_test.php: call to mod_assign_testable_assign->testable_view_batch_set_workflow_state()
* line 1617 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to mod_assign\event\events_test->test_batch_set_workflow_state_viewed()
* line 1223 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to PHPUnit\Framework\TestCase->runTest()
* line 72 of /lib/phpunit/classes/advanced_testcase.php: call to PHPUnit\Framework\TestCase->runBare()
* line 729 of /vendor/phpunit/phpunit/src/Framework/TestResult.php: call to advanced_testcase->runBare()
* line 973 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to PHPUnit\Framework\TestResult->run()
* line 685 of /vendor/phpunit/phpunit/src/Framework/TestSuite.php: call to PHPUnit\Framework\TestCase->run()
* line 651 of /vendor/phpunit/phpunit/src/TextUI/TestRunner.php: call to PHPUnit\Framework\TestSuite->run()
* line 146 of /vendor/phpunit/phpunit/src/TextUI/Command.php: call to PHPUnit\TextUI\TestRunner->run()
* line 99 of /vendor/phpunit/phpunit/src/TextUI/Command.php: call to PHPUnit\TextUI\Command->run()
* line 107 of /vendor/phpunit/phpunit/phpunit: call to PHPUnit\TextUI\Command::main()
* line 122 of /vendor/bin/phpunit: call to include()
Debugging: Callback before_standard_top_of_body_html in tool_ribbons component should be migrated to new hook callback for core\hook\output\before_standard_top_of_body_html_generation
* line 7479 of /lib/moodlelib.php: call to debugging()
* line 7593 of /lib/moodlelib.php: call to {closure}()
* line 93 of /lib/classes/hook/output/before_standard_top_of_body_html_generation.php: call to get_plugins_with_function()
* line 832 of /lib/outputrenderers.php: call to core\hook\output\before_standard_top_of_body_html_generation->process_legacy_callbacks()
* line 219 of /lib/mustache/src/Mustache/Context.php: call to core_renderer->standard_top_of_body_html()
* line 138 of /lib/mustache/src/Mustache/Context.php: call to Mustache_Context->findVariableInStack()
* line 30 of udata/localcache/mustache/1/boost/__Mustache_4aa1b29b14b62e9eb02cc88038b79589.php: call to Mustache_Context->findDot()
* line 67 of /lib/mustache/src/Mustache/Template.php: call to __Mustache_4aa1b29b14b62e9eb02cc88038b79589->renderInternal()
* line 216 of /lib/outputrenderers.php: call to Mustache_Template->render()
* line 107 of /theme/boost/layout/drawers.php: call to renderer_base->render_from_template()
* line 1483 of /lib/outputrenderers.php: call to include()
* line 1408 of /lib/outputrenderers.php: call to core_renderer->render_page_layout()
* line 273 of /mod/assign/classes/output/renderer.php: call to core_renderer->header()
* line 512 of /lib/outputrenderers.php: call to mod_assign\output\renderer->render_assign_header()
* line 5131 of /mod/assign/locallib.php: call to plugin_renderer_base->render()
* line 121 of /mod/assign/tests/fixtures/testable_assign.php: call to assign->view_batch_set_workflow_state()
* line 1035 of /mod/assign/tests/event/events_test.php: call to mod_assign_testable_assign->testable_view_batch_set_workflow_state()
* line 1617 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to mod_assign\event\events_test->test_batch_set_workflow_state_viewed()
* line 1223 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to PHPUnit\Framework\TestCase->runTest()
* line 72 of /lib/phpunit/classes/advanced_testcase.php: call to PHPUnit\Framework\TestCase->runBare()
* line 729 of /vendor/phpunit/phpunit/src/Framework/TestResult.php: call to advanced_testcase->runBare()
* line 973 of /vendor/phpunit/phpunit/src/Framework/TestCase.php: call to PHPUnit\Framework\TestResult->run()
* line 685 of /vendor/phpunit/phpunit/src/Framework/TestSuite.php: call to PHPUnit\Framework\TestCase->run()
* line 651 of /vendor/phpunit/phpunit/src/TextUI/TestRunner.php: call to PHPUnit\Framework\TestSuite->run()
* line 146 of /vendor/phpunit/phpunit/src/TextUI/Command.php: call to PHPUnit\TextUI\TestRunner->run()
* line 99 of /vendor/phpunit/phpunit/src/TextUI/Command.php: call to PHPUnit\TextUI\Command->run()
* line 107 of /vendor/phpunit/phpunit/phpunit: call to PHPUnit\TextUI\Command::main()
* line 122 of /vendor/bin/phpunit: call to include()


/home/stefan/dev/projects/moodle44/lib/phpunit/classes/advanced_testcase.php:97

ERRORS!
Tests: 1, Assertions: 2, Errors: 1.
```

I'll submit a fix for this momentarily.

--- a/classes/hook_callbacks.php
+++ b/classes/hook_callbacks.php
@@ -1,0 +1,56 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Hook callbacks for tool_ribbons.
+ * @package    tool_ribbons
+ * @copyright  2020 onwards Conn Warwicker
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_ribbons;
+
+use core\hook\output\before_standard_head_html_generation;
+use core\hook\output\before_standard_top_of_body_html_generation;
+
+defined('MOODLE_INTERNAL') || die;
+
+/**
+ * Hook callbacks for tool_ribbons.
+ * @package    tool_ribbons
+ * @copyright  2020 onwards Conn Warwicker
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class hook_callbacks {
+
+    /**
+     * Adds the markup for the ribbons to the top of the page body.
+     *
+     * @param \core\hook\output\before_standard_top_of_body_html_generation $hook
+     */
+    public static function before_standard_top_of_body_html(before_standard_top_of_body_html_generation $hook): void {
+        tool_ribbons_before_standard_top_of_body_html_callback($hook);
+    }
+
+    /**
+     * Adds the CSS for the ribbons in a <style> tag to the page header.
+     *
+     * @param \core\hook\output\before_standard_head_html_generation $hook
+     */
+    public static function before_standard_head_html(before_standard_head_html_generation $hook): void {
+        tool_ribbons_before_standard_html_head_callback($hook);
+    }
+}

--- a/db/hooks.php
+++ b/db/hooks.php
@@ -1,0 +1,41 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Hook callbacks for tool_ribbons
+ *
+ * @package    tool_ribbons
+ * @copyright  2020 Conn Warwicker
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use core\hook\output\before_standard_head_html_generation;
+use core\hook\output\before_standard_top_of_body_html_generation;
+
+defined('MOODLE_INTERNAL') || die();
+
+$callbacks = [
+    [
+        'hook' => before_standard_top_of_body_html_generation::class,
+        'callback' => '\tool_ribbons\hook_callbacks::before_standard_top_of_body_html',
+        'priority' => 0,
+    ],
+    [
+        'hook' => before_standard_head_html_generation::class,
+        'callback' => '\tool_ribbons\hook_callbacks::before_standard_head_html',
+        'priority' => 0,
+    ],
+];

--- a/lib.php
+++ b/lib.php
@@ -26,11 +26,19 @@ use tool_ribbons\ribbon;
 defined('MOODLE_INTERNAL') || die();
 
 /**
- * Generate HTML to display the ribbons.
- * @return string
+ * Callback to add HTML content to the top of the page body.
+ * Here, we use it to add the markup for the ribbons.
+ *
+ * This function is implemented here and used from two locations:
+ * - tool_ribbons_before_standard_top_of_body_html() in lib.php (for releases up to Moodle 4.3)
+ * - tool_ribbons\hook_callbacks::before_standard_top_of_body_html() in classes/hook_callbacks.php (for Moodle 4.4+).
+ *
+ * @param \core\hook\output\before_standard_top_of_body_html_generation $hook If the hook is passed, then the hook
+ *                                                                            implementation will be used. If not, then
+ *                                                                            legacy implementation will be used.
+ * @return string|void The legacy implementation returns a string, the hook implementation returns nothing.
  */
-function tool_ribbons_before_standard_top_of_body_html() : string {
-
+function tool_ribbons_before_standard_top_of_body_html_callback(&$hook = null) {
     $output = '';
 
     // Load the ribbons.
@@ -41,16 +49,30 @@ function tool_ribbons_before_standard_top_of_body_html() : string {
         $output .= $ribbon->display();
     }
 
-    return $output;
-
+    if ($hook) {
+        // Pass the markup to the hook.
+        $hook->add_html($output);
+    } else {
+        // Return the markup.
+        return $output;
+    }
 }
 
-/**
- * Generate CSS to add to the page to style the ribbons.
- * @return string
- */
-function tool_ribbons_before_standard_html_head() : string {
 
+/**
+ * Callback to add elements to the page <head> html tag.
+ * Here, we use it to add the CSS for the ribbons in a <style> to the page header.
+ *
+ * This function is implemented here and used from two locations:
+ * - tool_ribbons_before_standard_html_head() in lib.php (for releases up to Moodle 4.3)
+ * - tool_ribbons\hook_callbacks::before_standard_head_html_generation() in classes/hook_callbacks.php (for Moodle 4.4+).
+ *
+ * @param \core\hook\output\before_standard_head_html_generation $hook If the hook is passed, then the hook
+ *                                                                            implementation will be used. If not, then
+ *                                                                            legacy implementation will be used.
+ * @return string|void The legacy implementation returns a string, the hook implementation returns nothing.
+ */
+function tool_ribbons_before_standard_html_head_callback(&$hook = null) {
     $output = '<style type="text/css">';
 
     // Load the ribbons.
@@ -63,6 +85,28 @@ function tool_ribbons_before_standard_html_head() : string {
 
     $output .= '</style>';
 
-    return $output;
+    if ($hook) {
+        // Pass the markup to the hook.
+        $hook->add_html($output);
+    } else {
+        // Return the markup.
+        return $output;
+    }
+}
 
+
+/**
+ * Legacy callback that adds the markup for the ribbons to the top of the page body.
+ * @return string
+ */
+function tool_ribbons_before_standard_top_of_body_html() : string {
+    return tool_ribbons_before_standard_top_of_body_html_callback();
+}
+
+/**
+ * Legacy callback that adds CSS for the ribbons in a <style> tag to the page header.
+ * @return string
+ */
+function tool_ribbons_before_standard_html_head() : string {
+    return tool_ribbons_before_standard_html_head_callback();
 }

